binlog_driver/reader/reader.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reader
+
+import (
+	"github.com/Shopify/sarama"
+	"github.com/ngaut/log"
+	pb "github.com/pingcap/tidb-tools/binlog_proto/go-binlog"
+)
+
+func init() {
+	// log.SetLevel(log.LOG_LEVEL_NONE)
+	sarama.MaxResponseSize = 1 << 30
+}
+
+// Config for Reader
+type Config struct {
+	KafakaAddr []string
+	// if ConmmitTs is specified, reader will start return the binlog which the ts bigger than the config ts
+	CommitTS  int64
+	Offset    int64 // start at kafka offset
+	ClusterID string
+}
+
+// Message read from reader
+type Message struct {
+	Binlog *pb.Binlog
+	Offset int64 // kafka offset
+}
+
+// Reader to read binlog from kafka
+type Reader struct {
+	cfg    *Config
+	client sarama.Client
+
+	msgs      chan *Message
+	stop      chan struct{}
+	clusterID string
+}
+
+func (r *Reader) getTopic() string {
+	return r.cfg.ClusterID + "_obinlog"
+}
+
+// NewReader return a instance of Reader
+func NewReader(cfg *Config) (r *Reader, err error) {
+	r = &Reader{
+		cfg:       cfg,
+		stop:      make(chan struct{}),
+		msgs:      make(chan *Message, 1024),
+		clusterID: cfg.ClusterID,
+	}
+
+	r.client, err = sarama.NewClient(r.cfg.KafakaAddr, nil)
+	if err != nil {
+		r = nil
+		return
+	}
+
+	if r.cfg.CommitTS > 0 {
+		r.cfg.Offset, err = r.getOffsetByTS(r.cfg.CommitTS)
+		if err != nil {
+			r = nil
+			return
+		}
+		log.Debug("set offset to: ", r.cfg.Offset)
+	}
+
+	go r.run()
+
+	return
+}
+
+// Close shuts down the reader
+func (r *Reader) Close() {
+	close(r.stop)
```

Comment:
The `r.msgs`?

Reply:
No need to wait utill `run grouting` exists? gorutine leak and lifetime management

Reply:
close r.msgs at line L139 now, when you stop,  `run gorouting` will quit and close r.msg,  no need to wait run quit before return here

Reply:
yep, u can do it like this. But there is a real risk of a goroutineleak, because the select select select branch with some randomness, in addition user can close and new reader many times. Anyway, as u like